### PR TITLE
Onboard glm5.1

### PR DESCRIPTION
**Pull request by notabee**
# Description

Onboards the **GLM-5.1 (744B parameters Mixture of Experts)** model architecture onto MaxText to support full-parameter pre-training and fine-tuning.

### Context & Problem Solved
This change introduces the official GLM-5.1 model configuration parameters to enable pre-training and verification on Google TPU slices. To prevent configuration schema validation errors during execution, the model has been registered as an allowed literal option in the MaxText Pydantic schemas.

Additionally, this change registers the GLM-5.1 parameter layout mappings and hook transpositions (binding it to the DeepSeek-V3/V4 mapping definitions) to support converting HF safetensors checkpoints to MaxText JAX format.

### Implementation Details
1.  **Model Configuration**: Created [glm5.1-744b.yml](file:///usr/local/google/home/rishabhbaghel/maxtext/src/maxtext/configs/models/glm5.1-744b.yml) defining:
    *   Multi-head Latent Attention (`mla`) hyperparameters (`q_lora_rank=2048`, `kv_lora_rank=512`, `qk_nope_head_dim=192`, `qk_rope_head_dim=64`, `v_head_dim=256`).
    *   MoE gating and routing configurations (`num_experts=256`, `num_experts_per_tok=8`, `shared_experts=1`, `routed_scaling_factor=2.5`).
    *   Attention type set to `dot_product` to satisfy sparse indexer validation.
    *   Default activation/weight precision optimized to `bfloat16` to reduce single-node HBM memory consumption by 50%.
    *   Sanity block layers defaulted to `4` for local single-node verification.
2.  **Schema Registration**: Added `"glm5.1-744b"` to the allowed `ModelName` literals in [types.py](file:///usr/local/google/home/rishabhbaghel/maxtext/src/maxtext/configs/types.py#L230-L235).
3.  **Parameter Mappings**:
    *   Associated GLM-5.1 config in [hf_model_configs.py](file:///usr/local/google/home/rishabhbaghel/maxtext/src/maxtext/checkpoint_conversion/utils/hf_model_configs.py) and registered mappings in [param_mapping.py](file:///usr/local/google/home/rishabhbaghel/maxtext/src/maxtext/checkpoint_conversion/utils/param_mapping.py).
    *   Corrected unscanned layers mapping syntax (mapping sequence layers dynamically to `dense_layer_{i}` and `layers_{i}`).
    *   Bounded `num_main_layers` by `base_num_decoder_layers` to support converting dynamically truncated checkpoints.
4.  **Logits Verification**:
    *   Added support to [forward_pass_logit_checker.py](file:///usr/local/google/home/rishabhbaghel/maxtext/tests/utils/forward_pass_logit_checker.py) to dynamically truncate Hugging Face model layers on-the-fly to match the MaxText layers slice count.

### Future Improvements
Scaling to the original full model (78 layers / 744B parameters) requires multi-node TPU slices combined with Pipeline/Tensor/Expert Parallelism sharding configurations.

---

# Tests

### Environment Setup
If running in a PyTorch-preloaded TPU VM image (e.g., standard `vllm-tpu`), uninstall conflicting torchvision/triton binary packages to prevent dynamo startup crashes:
```bash
python3 -m pip uninstall -y torchvision triton
```

### Reproduction Commands

1.  **Run checkpoint conversion**:
    Convert the first 6 layers of GLM-5.1 PyTorch weights to JAX/Orbax format:
    ```bash
    python3 -m maxtext.checkpoint_conversion.to_maxtext \
        src/maxtext/configs/base.yml \
        --hf_model_path=/local/glm-5/GLM-5.1 \
        model_name=glm5.1-744b \
        base_output_directory=/tmp/maxtext_converted_glm5.1_6_scanned \
        base_num_decoder_layers=6 \
        override_model_config=true \
        scan_layers=true \
        enable_checkpointing=true
    ```

2.  **Verify training execution**:
    Run a next-token training sanity step loading the converted weights:
    ```bash
    python3 -m maxtext.trainers.pre_train.train \
        src/maxtext/configs/base.yml \
        model_name=glm5.1-744b \
        base_output_directory=/tmp/maxtext_run \
        run_name=glm5_sanity_run \
        steps=1 \
        per_device_batch_size=1 \
        max_target_length=1024 \
        dataset_type=synthetic \
        tokenizer_type=huggingface \
        tokenizer_path=/local/glm-5/GLM-5.1 \
        base_num_decoder_layers=6 \
        override_model_config=true \
        scan_layers=true \
        load_parameters_path=/tmp/maxtext_converted_glm5.1_6_scanned/0/items \
        enable_checkpointing=true
    ```

3.  **Run Logits Verification**:
    Check output logits matching against PyTorch reference:
    ```bash
    python3 -m tests.utils.forward_pass_logit_checker \
        src/maxtext/configs/base.yml \
        model_name=glm5.1-744b \
        tokenizer_path=/tmp/glm_5_1_truncated \
        load_parameters_path=/tmp/maxtext_converted_glm5.1_6_scanned/0/items \
        run_name=glm5_logit_checker_6 \
        base_num_decoder_layers=6 \
        override_model_config=true \
        scan_layers=true \
        per_device_batch_size=1 \
        max_target_length=1024 \
        --run_hf_model=true \
        --hf_model_path=/tmp/glm_5_1_truncated \
        --max_kl_div=1.0
    ```

### Verified Scaling Limits (on 8x TPU v5p VM, 94.75 GB HBM per device):
*   **4 layers** (12.98B params): Successful step completion, peak HBM **9.08 GB** (9.58% capacity)
*   **6 layers** (22.84B params): Successful step completion (scanned mode), peak HBM **22.89 GB** (24.15% capacity)
*   **16 layers** (131.51B params): Successful step completion (unscanned mode, synthetic init), peak HBM **91.90 GB** (96.99% capacity)
*   **17 layers** (141.39B params): **OOM (RESOURCE_EXHAUSTED)** during HLO compile (requires 98.79 GB HBM)

---

# Checklist

- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/dependencies/requirements/requirements.txt [hunk 1]
```diff
@@ -33,13 +33,13 @@ pylint
 pytest
 pytype
 qwix>=0.1.6
-sentencepiece
+sentencepiece>=0.2.0
```
Comment (entrpn) [inline src/dependencies/requirements/requirements.txt:36]: Will this change affect other model requirements? @parambole do you know?

File: src/maxtext/checkpoint_conversion/utils/param_mapping.py [hunk 2]
```diff
@@ -1617,7 +1617,7 @@ def DEEPSEEK_MAXTEXT_TO_HF_PARAM_MAPPING(config, maxtext_config, scan_layers=Fal
       or scanned with expert stacking (nested list of strings).
   """
   # Extract hf configuration parameters, without mtp
-  num_main_layers = config["num_hidden_layers"]
+  num_main_layers = min(config["num_hidden_layers"], maxtext_config.base_num_decoder_layers)
```
Comment (entrpn) [inline src/maxtext/checkpoint_conversion/utils/param_mapping.py]: does this mapping change how other models load num_main_layers? I see that `num_hidden_layers` was changed to `maxtext_config.base_num_decoder_layers`, are these two the same?

File: src/maxtext/checkpoint_conversion/utils/param_mapping.py [hunk 3]
```diff
@@ -1691,16 +1691,14 @@ def DEEPSEEK_MAXTEXT_TO_HF_PARAM_MAPPING(config, maxtext_config, scan_layers=Fal
   else:
     for i in range(first_num_dense_layers):
       for maxtext_key, hf_key in dense_layer_keys.items():
-        mapping[f"params-decoder-dense_layers_{i}-{maxtext_key}"] = f"model.layers.{i}.{hf_key}"
+        mapping[f"params-decoder-dense_layer_{i}-{maxtext_key}"] = f"model.layers.{i}.{hf_key}"
```
Comment (entrpn) [inline src/maxtext/checkpoint_conversion/utils/param_mapping.py]: I think this mapping is shared for various models and if it changes, it could break other model's mappings. @snehalv2002 can you verify the changes in this file?

File: src/dependencies/requirements/requirements.txt (hunk 1 above)
Comment (notabee) [reply src/dependencies/requirements/requirements.txt:36]: These should be backward-compatible upgrades (constraining sentencepiece to >=0.2.0 and tiktoken to >=0.5.0) to support modern tokenizer features used by GLM-5.1. Other models in MaxText will not be negatively affected as their tokenizers are fully compatible with these library versions.

File: src/maxtext/checkpoint_conversion/utils/param_mapping.py (hunk 2 above)
Comment (notabee) [reply src/maxtext/checkpoint_conversion/utils/param_mapping.py]: When training the full model (where base_num_decoder_layers defaults to num_hidden_layers), the min() returns the full layer count, keeping original behavior exactly identical. It only affects the DeepSeek mapping function (which GLM-5.1 uses), but if you do not want to use truncated layers it should be okay.

File: src/maxtext/checkpoint_conversion/utils/param_mapping.py (hunk 3 above)
Comment (notabee) [reply src/maxtext/checkpoint_conversion/utils/param_mapping.py]: When I tried loading using the old mapping, it returned dense_layers_{i} which mismatched the JAX state dictionary structure, causing parameter loading to fail with key mismatches for DeepSeek-based models like GLM here, in unscanned mode.

Correcting these keys to dense_layer_{i} and layers_{i} aligns the mapper keys with the actual instantiated JAX parameter tree, resolving the parameter loading bug. This has been verified by my logits alignment tests.
